### PR TITLE
appium test for keeping public keys updated

### LINKED ISSUE #1301: ui test: app updates older public keys to newer but not vice versa
These two tests can be searched on browser extension test suite by name: 
 - `do not auto-refresh key if older version of the same key available on attester`
 - `auto-refresh expired key if newer version of the same key available`

The iOS test will be similar, except instead of manual key import, all keys will be imported from attester (iOS doesn't support manually importing public keys into the app).

Setup: add two more public keys in `attesterPublicKeySamples`:

`keyOlderVersion`:

```
-----BEGIN PGP PUBLIC KEY BLOCK-----

xjMEYd8pKhYJKwYBBAHaRw8BAQdAXQ9bBzlYPwy3mQD5MIQSkuOyEomESHHo
AAiUi0enB77NKFVwZGF0aW5nIGtleSA8dXBkYXRpbmcua2V5QGV4YW1wbGUu
dGVzdD7CeAQQFgoAIAUCYd8pKgYLCQcIAwIEFQgKAgQWAgEAAhkBAhsDAh4B
AAoJEMfJkMGi6njd9HUA/0ZdZaaOFy0hM+GpEnzK+A/G3bLe9Kulh4jT8+4j
JqcKAP0V+pga+B1v98aeF8cRlgQPEyWtYUqZLcDLBQ6r3BEfDM44BGHfKSoS
CisGAQQBl1UBBQEBB0BPkVSyVsZ+vsF4e4NbVsq/YNjqL0JQI+t6OHc5YxJS
SgMBCAfCYQQYFggACQUCYd8pKgIbDAAKCRDHyZDBoup43Y+gAQDSVD/EDqCE
dSL33ptMlhwRCKHGiKcVmKwucxYkk6apFQEA018565fZcvtb339L2s/IIxLs
4621FX8Sy6kpR7mAzQo=
=3UnZ
-----END PGP PUBLIC KEY BLOCK-----
```

`keyNewerVersion`:

```
-----BEGIN PGP PUBLIC KEY BLOCK-----

xjMEYd8pKhYJKwYBBAHaRw8BAQdAXQ9bBzlYPwy3mQD5MIQSkuOyEomESHHo
AAiUi0enB77NKFVwZGF0aW5nIGtleSA8dXBkYXRpbmcua2V5QGV4YW1wbGUu
dGVzdD7CeAQQFgoAIAUCYd8pKgYLCQcIAwIEFQgKAgQWAgEAAhkBAhsDAh4B
AAoJEMfJkMGi6njd9HUA/0ZdZaaOFy0hM+GpEnzK+A/G3bLe9Kulh4jT8+4j
JqcKAP0V+pga+B1v98aeF8cRlgQPEyWtYUqZLcDLBQ6r3BEfDM44BGHfKSoS
CisGAQQBl1UBBQEBB0BPkVSyVsZ+vsF4e4NbVsq/YNjqL0JQI+t6OHc5YxJS
SgMBCAfCYQQYFggACQUCYd8pKgIbDAAKCRDHyZDBoup43Y+gAQDSVD/EDqCE
dSL33ptMlhwRCKHGiKcVmKwucxYkk6apFQEA018565fZcvtb339L2s/IIxLs
4621FX8Sy6kpR7mAzQrOOARh3yuYEgorBgEEAZdVAQUBAQdAglzBCJCRj29J
THYvVGaNESiiVKmyrTEnXonGUS58TwMDAQgHwmEEGBYIAAkFAmHfK5gCGwwA
CgkQx8mQwaLqeN3PWwD9ErvC+ufnX0O2AmZDz67QfFH6tA1t1/wUEHgzBXEe
gc8BAMaYm3AlSGbX1rJYgUtCWukkLuURdECIzerG2UuP87ID
=dQen
-----END PGP PUBLIC KEY BLOCK-----
```

Both can be written as single test called `app updates older public keys to newer but not vice versa` as follows:

stage 1

1) serve `keyOlderVersion` on attester mock for email `updating.key@example.test`
2) compose new email, enter `updating.key@example.test`, wait for it to load
3) go to Settings -> Contacts -> `updating.key@example.test` -> public keys (tap the first one)
  - Check the date of the "signature" - expect that exact string.
  - check "last fetched" and remember it to compare in below steps
  - check fingerprints - there should be exactly two of them - expect exactly these two

stage 2 

4) change attester mock settings to serve `keyNewerVersion` for `updating.key@example.test`
5) compose new email, enter `updating.key@example.test`, wait for it to load
6) go to Settings -> Contacts -> `updating.key@example.test` -> public keys (tap the first one)
  - Check the date of the "signature" - this should now be different, more recent time than in stage 1. Expect this exact (different then before) string in the test
  - check "last fetched" and confirm that it's a more recent time than stage 1
  - check fingerprints - there should be exactly three of them - expect exactly these three at this stage 2

stage 3

7) change attester mock settings to serve `keyOlderVersion` for `updating.key@example.test` again
8) compose new email, enter `updating.key@example.test`, wait for it to load
9) go to Settings -> Contacts -> `updating.key@example.test` -> public keys (tap the first one)
  - Check the date of the "signature" - this should remain the same as stage 2
  - check "last fetched" and confirm that it's a more recent time, even more recent than stage 2
  - check fingerprints - there should be exactly three of them - expect exactly these three at this step

stage 4:

10) delete the public key by tapping the trash icon on that screen
11) compose new email, enter `updating.key@example.test`, wait for it to load
12) go to Settings -> Contacts -> `updating.key@example.test` -> public keys (tap the first one)
  - Check the date of the "signature" - expect exactly the same as in stage 1
  - check "last fetched" - the date/time should be later than all previous stages
  - check fingerprints - expect exactly the same as stage 1

### PR DESCRIPTION
**Pull request by fcvakintos**
…ce versa"

This PR contains new tests for "app updates older public keys to newer", added new screen for pubkey details, added new method for datahelper - convertDateToMsec, updated tests

close #1301 

----------------------------------

**Tests** 
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+          clientConfiguration: {
+              flags: ["NO_PRV_CREATE", "NO_PRV_BACKUP", "NO_ATTESTER_SUBMIT", "PRV_AUTOIMPORT_OR_AUTOGEN", "FORBID_STORING_PASS_PHRASE"],
+              key_manager_url: "https://ekm.flowcrypt.com",
+          }
+      };
+      mockApi.attesterConfig = {
+          servedPubkeys: {
+              'updating.key@example.test': attesterPublicKeySamples.keyOlderVersion
+          }
+      };
```
Comment (sosnovsky) [inline appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:36]: Looks like indentation is a bit off on lines 22-36, they should be moved 1 tab to the left.

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      await ContactScreen.clickOnContact(userEmail);
+      await ContactPublicKeyScreen.checkPgpUserId(userEmail);
+      await ContactPublicKeyScreen.checkPublicKeyDetailsNotEmpty();
+      await ContactPublicKeyScreen.clickOnFingerPrint();
+
+      await PublicKeyDetailsScreen.checkPublicKeyDetailsScreen();
+      await PublicKeyDetailsScreen.checkPublicKeyNotEmpty();
+      await PublicKeyDetailsScreen.checkSignatureDateValue(newSignatureDate);
+      secondFetchedDate = await DataHelper.convertDateToMSec(await PublicKeyDetailsScreen.getLastFetchedDateValue());
+      await PublicKeyDetailsScreen.checkFingerPrintsValue(newFingerprintsValue);
```
Comment (sosnovsky) [inline appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:117]: Is it possible to have helper functions to remove code duplication?

As currently checks on lines 47-73, 91-117, 137-163 are identical. It'll be great to have, for example, `checkSignatureAndFingerprints(userEmail, signatureDate, fingerprintsValue)` function which will perform all these checks. It'll make test more readable

Comment (tomholub) [reply appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:117]: agree

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts (hunk 1 above)
Comment (fcvakintos) [reply appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:36]: changed

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts (hunk 2 above)
Comment (fcvakintos) [reply appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts]: done

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts [hunk 3]
```diff
@@ -89,32 +66,9 @@ describe('SETUP: ', () => {
...
-      await SettingsScreen.clickOnSettingItem('Contacts');
-
-      await ContactScreen.checkContactScreen();
-      await ContactScreen.checkContact(userEmail);
-      // Go to Contact screen
-      await ContactScreen.clickOnContact(userEmail);
-      await ContactPublicKeyScreen.checkPgpUserId(userEmail);
-      await ContactPublicKeyScreen.checkPublicKeyDetailsNotEmpty();
-      await ContactPublicKeyScreen.clickOnFingerPrint();
+      await PublicKeyHelper.checkSignatureAndFingerprints(userEmail, newSignatureDate, newFingerprintsValue);
```
Comment (sosnovsky) [inline appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:69]: `await MailFolderScreen.checkInboxScreen()` should be included in `checkSignatureAndFingerprints` too, as it's performed each time

**Review by sosnovsky — CHANGES_REQUESTED** [review]
Looks good now, just 1 small change needed

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts (hunk 3 above)
Comment (fcvakintos) [reply appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:66]: ok, pushed

File: appium/tests/data/index.ts [hunk 4]
```diff
@@ -53,6 +53,15 @@ export const CommonData = {
   longEmail: {
     message: '\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nText'
   },
+  updateUser: {
```
Comment (tomholub) [inline appium/tests/data/index.ts:56]: `updateRecipientPublicKey`

File: appium/tests/data/index.ts [hunk 5]
```diff
@@ -53,6 +53,15 @@ export const CommonData = {
...
+    email: 'updating.key@example.test',
+    oldSignatureDate: 'Jan 12, 2022 at 11:16:58 AM',//time for CI runs
+    newSignatureDate:  'Jan 12, 2022 at 11:27:20 AM',//time for CI runs
+    // oldSignatureDate: 'Jan 12, 2022 at 9:16:58 PM', //- for local run
+    // newSignatureDate:  'Jan 12, 2022 at 9:27:20 PM', // for local run
```
Comment (tomholub) [inline appium/tests/data/index.ts:61]: That's not good. Can you make sure that tests always run the iOS app on UTC time, both locally and in CI?

File: appium/tests/screenobjects/old-version-app.screen.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ import ElementHelper from "../helpers/ElementHelper";
 const SELECTORS = {
   BACK_BUTTON: '~arrow left c',
   SENDER_EMAIL: '~messageSenderLabel',
+  PUBLIC_KEY: '-ios class chain:**/XCUIElementTypeOther/XCUIElementTypeStaticText[2]'
```
Comment (tomholub) [inline appium/tests/screenobjects/old-version-app.screen.ts:7]: should have aid for this - can file a separate issue

**Review by sosnovsky — DISMISSED** [review]
👍

File: appium/tests/helpers/PublicKeyHelper.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import MailFolderScreen from "../screenobjects/mail-folder.screen";
+import NewMessageScreen from "../screenobjects/new-message.screen";
+import MenuBarScreen from "../screenobjects/menu-bar.screen";
+import SettingsScreen from "../screenobjects/settings.screen";
+import ContactScreen from "../screenobjects/contacts.screen";
+import ContactPublicKeyScreen from "../screenobjects/contact-public-key.screen";
+import PublicKeyDetailsScreen from "../screenobjects/public-key-details.screen";
+
+class PublicKeyHelper {
+ static checkSignatureAndFingerprints = async (userEmail: string, signatureDate: string , fingerprintsValue: string ) => {
```
Comment (tomholub) [inline appium/tests/helpers/PublicKeyHelper.ts:10]: should be called `loadRecipientInComposeThenCheckSignatureAndFingerprints` :shrug: 

Or maybe could be split into two methods. `loadRecipientInCompose` and `checkSignatureAndFingerprints`

File: appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      await ContactPublicKeyScreen.clickTrashButton();
+      await ContactScreen.checkContactScreen();
+      await ContactScreen.checkEmptyList();
+
+      await ContactScreen.clickBackButton();
+      await SettingsScreen.checkSettingsScreen();
+
+      await MenuBarScreen.clickMenuIcon();
+      await MenuBarScreen.clickInboxButton();
+
```
Comment (tomholub) [inline appium/tests/specs/mock/setup/UpdateOlderPubKeyToNewer.spec.ts:104]: Missing `// stage 4` here

**Review by tomholub — COMMENTED** [review]
Improvements, can be either here or in another PR.

Else looks good.

File: appium/tests/screenobjects/old-version-app.screen.ts (hunk 6 above)
Comment (fcvakintos) [reply appium/tests/screenobjects/old-version-app.screen.ts:7]: In old app we dont have aid for this element

Comment (tomholub) [reply appium/tests/screenobjects/old-version-app.screen.ts:7]: We can't be restricted by that. The old app should use the old selector, but we should use the new selector for the new app.

File: appium/tests/data/index.ts (hunk 4 above)
Comment (fcvakintos) [reply appium/tests/data/index.ts:56]: ok will change

File: appium/tests/data/index.ts (hunk 5 above)
Comment (fcvakintos) [reply appium/tests/data/index.ts:61]: i think it is a timezone issue

File: appium/tests/helpers/PublicKeyHelper.ts (hunk 7 above)
Comment (fcvakintos) [reply appium/tests/helpers/PublicKeyHelper.ts:10]: renamed to loadRecipientInComposeThenCheckSignatureAndFingerprints

File: appium/tests/screenobjects/old-version-app.screen.ts (hunk 6 above)
Comment (fcvakintos) [reply appium/tests/screenobjects/old-version-app.screen.ts:7]: for new app we use new selector
check `PublicKeyDetailsScreen`
Screenshot 2022-01-21 at 17 15 51

File: appium/tests/data/index.ts (hunk 5 above)
Comment (tomholub) [reply appium/tests/data/index.ts:61]: Of course it is a timezone issue. Can you set in appium runner to run the iOS simulator on UTC timezone regardless of the timezone of the host? There must be settings for it.

**Review by tomholub — APPROVED** [review]

**Pull request merged**